### PR TITLE
Neural Network Backpropagation and DQN Framework for training RL

### PR DESCRIPTION
**Pull request by SchindlerTo**
**Migrated from Bitbucket**
- Original Author: **Florian Rosner** *(no GitHub account)*
- Original Created: June 07, 2023 at 02:06 PM UTC
- Original URL: https://bitbucket.org/ultrazohm/ultrazohm_sw/pull-requests/390

---

#### Summary: What kind of change does this PR introduce?

Adds backpropagation and gradient descent to uz\_nn. Adds DQN Framework.

#### What is the current behavior?

uz\_nn can only be calculated feedforward, not trainable on the uz. There is no framework for Training DQN.

#### What is the new behavior?

Neural networks can be trained in the PS, with gradient descent and vector input/output and minibatch. Experience Buffer for DQN, DQN which contains critic and target net.

#### Does this PR introduce a breaking change?

Changes only software and tests for uz\_nn and matrix, adds new modules, not affecting vivado.

#### Which tests have to be done by reviewers before approving?

Implement a new network and train it and compare it with the Matlab code.

#### Other information:

Different functions were added, especially for uz\_matrix and uz\_nn, check for asserts

#### Please check if the PR fulfills these requirements

* Build pipelines pass
* Tests for the changes have been defined
* Docs have been added/updated

‌

### CODE REVIEW

**Comment by SchindlerTo** [comment]
Florian Rosner opened the pull request on June 07, 2023 at 02:06 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: adb0240ce09f by Florian Rosner on June 12, 2023 at 09:28 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 1b419dc6f955 by Florian Rosner on June 20, 2023 at 09:31 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: ef94fa33bf74 by Florian Rosner on June 27, 2023 at 03:28 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: e5b6767c520d by Florian Rosner on June 29, 2023 at 11:48 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: d98b06fa5ad5 by Florian Rosner on June 29, 2023 at 11:51 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 29d4293f8325 by Florian Rosner on July 10, 2023 at 01:11 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 7d23b47109c1 by Florian Rosner on July 18, 2023 at 01:11 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 5c7ce258611e by Florian Rosner on July 25, 2023 at 12:21 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: e40b9570e9c1 by Florian Rosner on July 31, 2023 at 12:43 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 16ce3298d891 by Florian Rosner on August 04, 2023 at 12:50 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: f374618cfa93 by Florian Rosner on August 08, 2023 at 12:42 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: ea141bcd253d by Florian Rosner on August 09, 2023 at 10:04 AM UTC

**Comment by SchindlerTo** [comment]
PR updated by Florian Rosner on August 09, 2023 at 10:09 AM UTC:
- Title updated from 'Neural Network Backpropagation' to 'Neural Network Backpropagation and DQN Framework for training RL'
- Description updated

**Comment by SchindlerTo** [comment]
New commit added to PR: ef1188751045 by Florian Rosner on August 09, 2023 at 11:04 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: b1e1dccab724 by Florian Rosner on August 14, 2023 at 01:08 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 3ee8f56088f6 by Florian Rosner on August 15, 2023 at 08:40 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 00b19703811d by Florian Rosner on August 16, 2023 at 01:03 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 96210d06f246 by Florian Rosner on August 21, 2023 at 09:07 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: f9c1737baa7b by Florian Rosner on August 22, 2023 at 01:56 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: f26c05344c31 by Florian Rosner on August 23, 2023 at 02:04 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: f46ee9977768 by Florian Rosner on August 25, 2023 at 12:02 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 200d6e210934 by Florian Rosner on August 29, 2023 at 12:47 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 78ede481bd51 by Florian Rosner on September 04, 2023 at 12:52 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 91d7c375ce1b by Florian Rosner on September 07, 2023 at 10:49 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 7fa28f71e47c by Florian Rosner on September 13, 2023 at 04:16 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 79437741fbb2 by Florian Rosner on September 14, 2023 at 01:35 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 016672400d48 by Florian Rosner on September 14, 2023 at 02:51 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 258652a78c3a by Tobias on September 15, 2023 at 08:24 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 48b0be00e794 by Tobias on September 15, 2023 at 08:52 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: a483f15c5900 by Florian Rosner on September 15, 2023 at 09:06 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: e3127f5243ed by Tobias on September 15, 2023 at 11:49 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 8f81388ddf96 by Tobias on September 15, 2023 at 11:59 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 770cefec4382 by Florian Rosner on September 19, 2023 at 09:35 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 7cbf24b0ebfe by Florian Rosner on September 19, 2023 at 09:41 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: f6aec6d1201b by Florian Rosner on September 20, 2023 at 09:48 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: da3815456468 by Florian Rosner on September 20, 2023 at 01:53 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: d6da4d0136a8 by Florian Rosner on September 21, 2023 at 01:46 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: cd508f7a0522 by Florian Rosner on September 25, 2023 at 11:33 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: a4a4b4f130e5 by Florian Rosner on September 26, 2023 at 12:55 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: cd009da7a120 by Florian Rosner on September 27, 2023 at 03:02 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 542ae1d28d52 by Florian Rosner on September 29, 2023 at 11:24 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 3b72dd9ee3c3 by Florian Rosner on October 02, 2023 at 11:22 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 6043a49c2a4f by Florian Rosner on October 02, 2023 at 12:30 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: fa0ff4b4b498 by Florian Rosner on October 02, 2023 at 06:01 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 2a3e1bd07d6a by Florian Rosner on October 03, 2023 at 08:00 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 646aca067e25 by Florian Rosner on October 05, 2023 at 08:03 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: e76e8d7fa2c0 by Florian Rosner on October 06, 2023 at 11:01 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 865afea03cbb by Florian Rosner on October 06, 2023 at 01:19 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: fdab473cf891 by Tobias on October 07, 2023 at 01:18 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 7a13f03dde9d by Tobias on October 10, 2023 at 11:31 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: ccf7d8118a79 by Tobias on October 10, 2023 at 11:56 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 3ed8cd3b564e by Tobias on October 10, 2023 at 03:49 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: e647ea0e4651 by Tobias on October 11, 2023 at 11:56 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: d9980a457fc2 by Tobias on October 11, 2023 at 08:40 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: f2576191b420 by Tobias on October 11, 2023 at 08:56 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: efaa01210135 by Tobias on October 12, 2023 at 02:18 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 257041e764d2 by Tobias on October 12, 2023 at 03:20 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 25b354ec185a by Tobias on October 13, 2023 at 11:56 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 0a785d0af828 by Tobias on October 13, 2023 at 12:27 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: a7c69a595750 by Tobias on October 13, 2023 at 12:42 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 5be6b79ab353 by Tobias on October 16, 2023 at 03:36 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 707825480e4a by Tobias on October 16, 2023 at 04:18 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 5c587cc58c91 by Tobias on October 18, 2023 at 02:12 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: aad5fb9336ec by Tobias on October 25, 2023 at 02:06 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 7f08b94237f8 by Tobias on December 10, 2023 at 08:26 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 1f6e2cf890f4 by Tobias on December 10, 2023 at 09:27 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 5968123f152e by Tobias on December 11, 2023 at 09:59 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 7b722eba0cbc by Tobias on December 11, 2023 at 11:02 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 1beccb0671e6 by Tobias on December 11, 2023 at 03:42 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: a1f11b10111c by Tobias on December 11, 2023 at 04:10 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 2ec13e74c788 by Tobias on December 12, 2023 at 12:14 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: e19d04a4ef04 by Tobias on December 12, 2023 at 04:40 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 569ea711b8f6 by Tobias on December 13, 2023 at 08:51 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 133fccc03274 by Tobias on December 13, 2023 at 04:26 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: b3bd4c9cb9b2 by Tobias on December 13, 2023 at 05:13 PM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: 99def590923a by Tobias on December 15, 2023 at 07:15 AM UTC

**Comment by SchindlerTo** [comment]
New commit added to PR: ba0cd91e95f1 by Tobias on December 19, 2023 at 09:44 PM UTC
